### PR TITLE
[#656] Enable users to connect via Twilio Sms and Whatsapp

### LINKED ISSUE #656: Enable users to connect via WhatsApp/SMS
### Summary
To enhance overall experience, we should offer the ability to connect SMS and WhatsApp channels after the Twilio requirements have been met.

### Twilio Requirement

On Airy Core, it should be an option to connect a number for WhatsApp or SMS messaging. Before this is possible, adding a Twilio Auth Token should be required. Before the Twilio Auth Token is povided and a phone number is purchased, we should inform system users that they cannot connect WhatsApp or SMS channels without these requirements.

104598079-6d9bf880-5676-11eb-9ad9-f25cc8d232d7

**Note**: in relation to the image above, we cannot direct system users to `airy.conf`. Therefore, we should initially direct system users to the correct [documentation](https://github.com/airyhq/airy/blob/develop/docs/docs/api/endpoints/channels.md).

### SMS 
Once a phone number is bought via Twilio, it should be possible to connect that number for SMS messaging. In this instance, it should also be possible to enter an icon and name to differentiate between channels from an Admin aspect.
2 1 0 Connect SMS@2x
It should be possible for several phone numbers to be added within the SMS channel, and these numbers will all be visibly listed.
2 2 0 Connect SMS@2x
It should be an option to be able to edit the information (phone number, icon, and name) given for each phone number listed within the channel.
2 2 1 Connect SMS Edit@2x

### WhatsApp
Once a phone number is bought via Twilio, it should be possible to connect that number for WhatsApp messaging. In this instance, it should also be possible to enter an icon and name to differentiate between channels from an Admin aspect.
4 0 0 Connect Whatsapp@2x
It should be possible for several phone numbers to be added within the WhatsApp channel. It should also be possible for the same phone number to be used for more than one location, if desired.
4 1 0 Connect WhatsApp@2x

### PR DESCRIPTION
**Pull request by ogbeche77**
Fixes #656

### CODE REVIEW

File: frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx [hunk 1]
```diff
@@ -60,9 +57,9 @@ const SourceInfo = (props: SourceInfoProps) => {
                   );
                 })}
               </div>
-              <div className={styles.extraChannel} onClick={props.onSourceInfoClick}>
+              <div className={styles.extraChannel} onClick={() => props.onSourceInfoClick()}>
```
Comment (AitorAlgorta) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx:60]: we can use `props.onSourceInfoClick` and it will also work, with less render cycles.

File: frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx [hunk 2]
```diff
@@ -60,9 +57,9 @@ const SourceInfo = (props: SourceInfoProps) => {
...
                 {hasExtraChannels && (
-                  <LinkButton onClick={props.onSourceInfoClick}>
+                  <LinkButton onClick={() => props.onSourceInfoClick()}>
```
Comment (AitorAlgorta) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx:62]: same here

File: frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx [hunk 3]
```diff
@@ -71,7 +68,7 @@ const SourceInfo = (props: SourceInfoProps) => {
           </div>
 
           <div className={styles.channelButton}>
-            <button type="button" className={styles.addChannelButton} onClick={props.onAddChannelClick}>
+            <button type="button" className={styles.addChannelButton} onClick={() => props.onAddChannelClick()}>
```
Comment (AitorAlgorta) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx:71]: same here

File: frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioSmsConnect.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      <h1 className={styles.headline}>SMS</h1>
+      <LinkButton onClick={props.history.goBack} type="button">
+        <BackIcon className={styles.backIcon} />
+        Back
+      </LinkButton>
+
+      <SmsWhatsappForm
+        connectTwilioSms={connectTwilioSms}
+        twilioPhoneNumber="Twilio Phone Number"
+        placeholder="Purchased Number +158129485394"
```
Comment (AitorAlgorta) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioSmsConnect.tsx:85]: placeholder shouldn't have a real number

File: frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioSmsConnect.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        twilioNumberInput={smsNumberInput}
+        handleNumberInput={handleNumberInput}
+        imageUrl="Image URL (optional)"
+        urlPlaceholder="Add an URL"
+        urlName="url"
+        urlText="url"
+        twilioUrlInput={smsUrlInput}
+        handleUrlInput={handleUrlInput}
+        accountName="Add a Name (optional)"
+        namePlaceholder="SMS Acme Berlin"
```
Comment (AitorAlgorta) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioSmsConnect.tsx:97]: placeholder shouldn't have a real name

File: frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioSmsSource.tsx [hunk 6]
```diff
@@ -1,34 +1,67 @@
...
+          }}
+          onChannelClick={(channel: Channel) => {
+            props.history.push(CHANNELS_TWILIO_SMS_ROUTE + `/${channel.id}`);
+          }}
+          onAddChannelClick={() => {
+            props.history.push(CHANNELS_TWILIO_SMS_ROUTE + '/new_account');
+          }}
+        />
+      </div>
+      {showModal && (
```
Comment (AitorAlgorta) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioSmsSource.tsx:55]: why don't show the modal that @thorstenairy built

**Review by AitorAlgorta — APPROVED** [review]

**Review by AitorAlgorta — CHANGES_REQUESTED** [review]
Left some comments

File: frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelsList.tsx [hunk 7]
```diff
@@ -1,25 +1,28 @@
...
+import {Channel, ChannelSource} from 'httpclient';
 import {allChannels} from './../../../selectors/channels';
-import {ChannelSource} from 'httpclient';
 import ChannelListItem from './ChannelListItem';
 import {SearchField} from '@airyhq/components';
 import styles from './ChannelsList.module.scss';
 
 type ChannelsListProps = {} & ConnectedProps<typeof connector> & RouteComponentProps<{source: string}>;
 
-const mapStateToProps = (state: StateModel, ownProps: RouteComponentProps<{source: string}>) => ({
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelsList.tsx:19]: I made this change yesterday to filter the channels inside the mapStateToProps. You should have accepted the incoming change here. :)

File: frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelsList.tsx [hunk 8]
```diff
@@ -32,9 +35,9 @@ const ChannelsList = (props: ChannelsListProps) => {
   const [showingSearchField, setShowingSearchField] = useState(false);
   const source = props.match.params.source;
 
-  const filteredChannels = channels.filter((channel: Channel) =>
-    channel?.metadata?.name?.toLowerCase().includes(searchText.toLowerCase())
-  );
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelsList.tsx:37]: same happened here :)

File: frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelsList.tsx [hunk 9]
```diff
@@ -74,6 +77,16 @@ const ChannelsList = (props: ChannelsListProps) => {
     }
   };
 
+  const channelSorter = (channelA: Channel, channelB: Channel) => {
+    if (channelA.metadata.name.toLowerCase() < channelB.metadata.name.toLowerCase()) {
+      return -1;
+    }
+    if (channelA.metadata.name.toLowerCase() > channelB.metadata.name.toLowerCase()) {
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelsList.tsx:84]: same happened here :) this function is removed and the sorting is happening inline with localeCompare()

File: frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx [hunk 10]
```diff
@@ -42,7 +39,7 @@ const SourceInfo = (props: SourceInfoProps) => {
                   return (
                     <li key={channel.sourceChannelId} className={styles.channelListEntry}>
                       <button className={styles.connectedChannelData} onClick={() => props.onChannelClick(channel)}>
-                        {channel.metadata?.imageUrl ? (
+                        {source === 'facebook' && channel.metadata?.imageUrl ? (
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/SourceInfo/index.tsx:42]: the additional check is not needed, cause if the channel doesn't have any imageUrl it's already false

File: frontend/ui/src/pages/Channels/index.tsx [hunk 11]
```diff
@@ -52,6 +55,9 @@ const Channels = (props: ChannelsConnectProps) => {
       <Route path={[`${CHANNELS_FACEBOOK_ROUTE}/:channelId?`]} component={FacebookConnect} />
       <Route path={[`${CHANNELS_CHAT_PLUGIN_ROUTE}/:channelId?`]} component={ChatPluginConnect} />
       <Route path={[`${CHANNELS_CONNECTED_ROUTE}/:source?`]} component={ChannelsList} />
+      <Route path={[`${CHANNELS_TWILIO_SMS_ROUTE}/:channelId?`]} component={TwilioSmsConnect} />
+      <Route path={[`${CHANNELS_CONNECTED_ROUTE}/:source?`]} component={ChannelsList} />
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/index.tsx:59]: ChannelsList is twice in here

File: frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioWhatsappConnect.tsx [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+  const [whatsappUrlInput, setWhatsappUrlInput] = useState('');
+  const channelId = props.match.params.channelId;
+
+  useEffect(() => {
+    if (channelId !== 'new_account') {
+      const channel = props.channels.find((channel: Channel) => {
+        return channel.id === channelId;
+      });
+      if (channel) {
+        setWhatsappNumberInput(channel.sourceChannelId || '');
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ChannelsMainPage/Sources/TwilioWhatsappConnect.tsx:36]: you can check this already for the initialState :)

File: frontend/ui/src/pages/Channels/Facebook/RequirementsDialog.tsx [hunk 13]
```diff
@@ -47,3 +47,51 @@ export const RequirementsDialog = (props: RequirementsDialogProps) => {
...
+          </button>
+          <div className={styles.inviteWrapper}>
+            <h1 className={styles.headline}>Connect with Twilio First</h1>
+            <p className={styles.firstMessage}>
+              Before you connect a number for SMS or Whatsapp, you must add a{' '}
+              <a
+                href="https://support.twilio.com/hc/en-us/articles/223136027-Auth-Tokens-and-How-to-Change-Them"
+                target="_blank"
+                rel="noreferrer">
+                Twilio Auth Token{' '}
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/Facebook/RequirementsDialog.tsx:67]: Nitpicking: this space extends the underscore by 1  so it isn't align with the phrase anymore

File: frontend/ui/src/pages/Channels/Facebook/RequirementsDialog.tsx [hunk 14]
```diff
@@ -47,3 +47,51 @@ export const RequirementsDialog = (props: RequirementsDialogProps) => {
...
+              </a>
+              to the{' '}
+              <a
+                className={styles.configMessage}
+                href="https://github.com/airyhq/airy/blob/develop/docs/docs/api/endpoints/channels.md"
+                target="_blank"
+                rel="noreferrer">
+                infrastructure/airy.conf{' '}
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/Facebook/RequirementsDialog.tsx:75]: same here

File: frontend/ui/src/pages/Channels/Facebook/RequirementsDialog.module.scss [hunk 15]
```diff
@@ -84,3 +84,34 @@
   cursor: pointer;
   outline: none;
 }
+.inviteWrapper {
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/Facebook/RequirementsDialog.module.scss:87]: as you added the Twilio requirementsDialog inside the Facebook/RequirementsDialog your css changes will affect the requirementsDialog of the facebook flow.

File: frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx [hunk 16]
```diff
@@ -27,32 +24,12 @@ const ChannelItem = (props: ChannelItemProps) => {
...
-      case ChannelSource.chatPlugin:
-        return <AiryLogo />;
-      default:
-        return <AiryLogo />;
-    }
+  const isPhoneNumberSource = () => {
+    return channel.source === 'twilio.sms' || channel.source === 'twilio.whatsapp';
   };
 
-  const ChannelIcon = ({channel}: {channel: Channel}) => {
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx:51]: why did you remove this? this Component is returning the default logo of the source or the given image of the imageUrl (as Facebook pages)

File: frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx [hunk 17]
```diff
@@ -74,18 +51,19 @@ const ChannelItem = (props: ChannelItemProps) => {
...
-            <div className={styles.channelName}>
-              {channel.metadata.name}
-              {channel.connected && (
-                <div className={styles.connectedHint}>
-                  Connected <CheckMark />
-                </div>
-              )}
+
+          <div className={styles.channelName}>{channel.metadata?.name}</div>
+          {isPhoneNumberSource() && <div className={styles.channelId}>{channel.sourceChannelId}</div>}
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx:59]: you need to check isPhoneNumberSource first and the return either channelName or channelId, otherwise it will have a gap where the name would have been. This might happen also in sourceInfo!

Comment (ogbeche77) [reply frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx:59]: For sms and whatsapp (UI design), we always need to return both channel name and channel id. I fixed the css, so it works for all sources. 
Screen Shot 2021-03-12 at 11 25 57
Screen Shot 2021-03-12 at 11 26 19

Comment (ogbeche77) [reply frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx:59]: I am not sure I understand the statement. If the user does not input a name, they don't get a name rendered in the UI. The current design looks like that. If name is not filled in the input bar, then the name filed will be blank, as it is optional
Screen Shot 2021-03-12 at 11 25 57

Comment (thorstenairy) [reply frontend/ui/src/pages/Channels/ConnectedChannelsList/ChannelListItem.tsx:59]: Yes the name is optional but on your screenshot you added always a name to it, if you don't add a name it looks like this:

Screenshot 2021-03-12 at 11 45 46

File: frontend/ui/src/pages/Channels/Twilio/SmsWhatsappForm.tsx [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+            placeholder={props.namePlaceholder}
+            value={props.twilioNameInput}
+            required={false}
+            height={32}
+            onChange={props.handleNameInput}
+            fontClass="font-s"
+          />
+        </div>
+        <Button type="submit" styleVariant="normal" disabled={props.twilioNumberInput.trim().length == 0}>
+          Connect SMS Number
```
Comment (thorstenairy) [inline frontend/ui/src/pages/Channels/Twilio/SmsWhatsappForm.tsx:67]: the title of the button should either display 'Connect SMS Number' or if you are updating the channel 'Update SMS Number'

**Review by AitorAlgorta — APPROVED** [review]
